### PR TITLE
Gen weight refactor main

### PR DESCRIPTION
**Pull request by dteague**
#### PR description:

Two commits aimed at fixing some bugs and adding some features.

First adds the "fillEmptyIfWeightFails" option. Still needs to be worked on. The only convenient spot I found to put it was in the GenWeightProductProducer. If that fails, it resizes the weight vector to 1 (aka, only the central weight) and continues on. This may be unintented and certainly can be expanded, but in my cursory look, this was the most obvious spot to place it.

Further, some errors were resolved. The CMSDriver commands and what is changed are explained below:

 `# with command line options: step1 --filein dbs:/ZZTo4L_TuneCP5_13TeV-amcatnloFXFX-pythia8/RunIIAutumn18MiniAOD-102X_upgrade2018_realistic_v15-v1/MINIAODSIM --fileout blah.root --mc --eventcontent NANOAODGEN --datatier NANOAODSIM --conditions auto:mc --step NANOGEN --nThreads 1 --no_exec --customise Configuration/DataProcessing/Utils.addMonitoring -n 1 --python_filename ZZTo4L_cfg.py`

This was the error mentioned by Sanghyun. The problem is conflict in how the xml file is cleaned: if the file is not consistent, it changes the actual header vector, if anything else is not working, it changes the string DERIVED from the header vector. This was causing conflict when the `isConsistent` function failed to see the xml wasn't consistent because of &lt; escape characters, so the escape characters were fixed to then the string that was fixed was replaced by the fixing of the consistency problem. 

That might not have made a ton of sense, but the take away is the Error handling was retooled a bit. It could definitely be done in a different way, but I opted for more modern code, ie `enum class` and if/else instead of case blocks

`# with command line options: step1 --filein dbs:/WWJTo2L2Nu_NNLOPS_TuneCP5_13TeV-powheg-pythia8/RunIIFall17MiniAODv2-PU2017_12Apr2018_94X_mc2017_realistic_v14-v1/MINIAODSIM --fileout blah.root --mc --eventcontent NANOAODGEN --datatier NANOAODSIM --conditions auto:mc --step NANOGEN --nThreads 1 --no_exec --customise Configuration/DataProcessing/Utils.addMonitoring -n 1 --python_filename WJ_cfg.py`

This file has weights not included in the header, namely NNLOP weights for random PDF weights, eg if PDF weight exists, there may be a NNLOP version of these weights that is not specified. The fix to the problem is to make sure that when adding an Unassociated Group that the product vector's size has to also be updated. Since this was causing the weight to fail though, it acted as a test for the `fillEmptyIfWeightFails` option.

I will be testing on more files, but these were files that straight up failed I've tested on so far

### CODE REVIEW

File: GeneratorInterface/Core/src/LHEWeightHelper.cc [hunk 1]
```diff
@@ -134,16 +134,16 @@ namespace gen {
     int close = -1;
     for (size_t idx = 0; idx < headerLines_.size(); idx++) {
       std::string line = headerLines_[idx];
-      if (line.find("<weightgroup") != std::string::npos) {
```
Comment (dteague) [inline GeneratorInterface/Core/src/LHEWeightHelper.cc:137]: This was the problem line for Sanghyun's file: if there are the xml escape characters, this code will think the xml is consistent because it will see NO weightgroups. Removing the < and > seems to work, though it may cause problems if there is a comment that contains "weightgroup". Might be safer to put this code as `line.find("<weightgroup") != std::string::npos || line.find("&lt;weightgroup") != std::string::npos`?

**Comment by dteague** [comment]
I'm testing out the files in [This sheet](https://docs.google.com/spreadsheets/d/1Qlg1c6qrMGY0q111Va2Vl0fJQUz15ZTjKEvRaV8rHaQ/edit#gid=0) and the emptyGenHeader still needs to be taken care of a an empty weight is needed so that is what I'm resolving now

**Comment by dteague** [comment]
The filling of the GenWeights with an empty product might work, but further debug needs to be done...
I was testing on 
/ggZH_HToBB_ZToLL_M120_13TeV_powheg_pythia8/RunIISummer16MiniAODv3-PUMoriond17_94X_mcRun2_asymptotic_v3-v1/MINIAODSIM
and it get to processing the events but it runs into an error seen before, namely:
```
----- Begin Fatal Exception 07-Jun-2021 12:16:54 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 2 stream: 0
   [1] Running path 'nanoAOD_step'
   [2] Calling method for module SimpleXYZPointFlatTableProducer/'genVertexTable'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: ROOT::Math::PositionVector3D<ROOT::Math::Cartesian3D<float>,ROOT::Math::DefaultCoordinateSystemTag>
Looking for module label: genParticles
Looking for productInstanceName: xyz0

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```

**Comment by dteague** [comment]
I've added a bit more to the code to fix problems that cropped up in the run so now all of the "Corner Cases" in the [Weight study tests sheet](https://docs.google.com/spreadsheets/d/1Qlg1c6qrMGY0q111Va2Vl0fJQUz15ZTjKEvRaV8rHaQ/edit#gid=0) run without fail, at least with the right flags turned on. More work will have to be done to make sure that the weights are indeed filled as expected, but the code now will process a file without crashing and looking at a couple of the output files, the weights look as one might expect.

**Comment by kdlong** [comment]
Sorry for being so slow. I'm looking now, trying to knock this out this week. I didn't look through everything, but it works and is an improvement, so I'm merging.
